### PR TITLE
feat: make getCurrentUserDetails() throw exception instead of returning NULL

### PR DESCRIPTION
**Pull request by netroms**
# Summary
The primary update is that getCurrentUserDetails() and getCurrentUsername() now throw exceptions when no user is logged in. In certain rare scenarios, such as when a user is not logged in yet in a public end-point, it’s recommended to first check using hasCurrentUser(). For background processes it's recommended to use the SystemUser as the authenticated user.

Previously, these methods returned NULL, which could indicate a background system process or be used as a convenient fallback when no user was set in the context, allowing tests to proceed as if they were run by a superuser. However, this introduced ambiguity in cases where NULL might represent a genuinely unauthenticated user, such as in public endpoints. This ambiguity increased code complexity by requiring frequent NULL checks and posed potential security risks due to unclear intent.

The `@Nonnull` annotations was also added to the methods, now making it clearer this is not intended to return NULL.
Some refactoring of the most low-hanging cases where `@Nonnull` is discovered by the static code analysis was refactored. There is still many cases left especially in the ACL related code, this will be addressed in follow up refactoring PRs.

## Automatic tests
CurrentUserUtilTest# testCurrentUsernameThrowsException
CurrentUserUtilTest# testCurrentUserDetailsThrowsException

## Manual tests
N/A

### CODE REVIEW

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/BaseIdentifiableObject.java [hunk 1]
```diff
@@ -420,8 +420,11 @@ public void setFavorites(Set<String> favorites) {
   @JsonProperty
   @JacksonXmlProperty(namespace = DxfNamespaces.DXF_2_0)
   public boolean isFavorite() {
+    if (!CurrentUserUtil.hasCurrentUser()) {
+      return false;
+    }
     UserDetails user = CurrentUserUtil.getCurrentUserDetails();
-    return user != null && favorites != null && favorites.contains(user.getUid());
+    return favorites != null && favorites.contains(user.getUid());
```
Comment (jbee) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/BaseIdentifiableObject.java:427]: The `favorites != null` part should be lifted up as a condition for early exit. In that case we do not even have to access the current user.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/user/CurrentUserUtil.java [hunk 2]
```diff
@@ -47,12 +47,13 @@ private CurrentUserUtil() {
...
    * @return the current user's username
    */
+  @Nonnull
   public static String getCurrentUsername() {
     Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
     if (authentication == null
         || !authentication.isAuthenticated()
         || authentication.getPrincipal() == null) {
-      return null;
+      throw new RuntimeException("No authenticated user found");
```
Comment (jbee) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/user/CurrentUserUtil.java:56]: maybe lets use `IllegalStateException` to be a bit more specific

File: dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/maintenance/DefaultMaintenanceService.java [hunk 3]
```diff
@@ -147,8 +147,8 @@ public void prunePeriods() {
   @Override
   @Transactional
   public boolean pruneData(OrganisationUnit organisationUnit) {
-    if (CurrentUserUtil.getCurrentUsername() == null
-        || !CurrentUserUtil.getCurrentUserDetails().isSuper()) {
+    CurrentUserUtil.getCurrentUsername();
```
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/maintenance/DefaultMaintenanceService.java:150]: I think this line can be removed now?

File: dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/maintenance/DefaultMaintenanceService.java [hunk 4]
```diff
@@ -166,8 +166,8 @@ public boolean pruneData(OrganisationUnit organisationUnit) {
   @Override
   @Transactional
   public boolean pruneData(DataElement dataElement) {
-    if (CurrentUserUtil.getCurrentUsername() == null
-        || !CurrentUserUtil.getCurrentUserDetails().isSuper()) {
+    CurrentUserUtil.getCurrentUsername();
```
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/maintenance/DefaultMaintenanceService.java:169]: same here

File: dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/event/data/AbstractAnalyticsServiceTest.java [hunk 5]
```diff
@@ -72,7 +73,7 @@
  * @author Luciano Fiandesio
  */
 @ExtendWith(MockitoExtension.class)
-class AbstractAnalyticsServiceTest {
+class AbstractAnalyticsServiceTest extends TestBase {
```
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/event/data/AbstractAnalyticsServiceTest.java:76]: I think it gets messy when mixing a mockito test with spring stuff. Maybe you can copy what you need instead or make a `MockitoTestBase`?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/dataapproval/hibernate/HibernateDataApprovalStore.java [hunk 6]
```diff
@@ -284,7 +284,7 @@
     User currentUser = userService.getUserByUsername(currentUsername);
 
     final String strArrayUserGroups =
-        (currentUser == null || CollectionUtils.isEmpty(currentUser.getGroups()))
+        (CollectionUtils.isEmpty(currentUser.getGroups()))
```
[the remark this answers is not in the document]
Comment (jbee) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/dataapproval/hibernate/HibernateDataApprovalStore.java:287]: Adding an annotation should now help with this?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/dataelement/DataElementDefaultDimensionPopulator.java [hunk 7]
```diff
@@ -107,9 +110,28 @@ public void executeInTransaction() {
...
+    context.setAuthentication(authentication);
+    SecurityContextHolder.setContext(context);
+  }
+
+  private static void clearContext() {
+    SecurityContext context = SecurityContextHolder.getContext();
+    if (context != null) {
+      SecurityContextHolder.getContext().setAuthentication(null);
+    }
+    SecurityContextHolder.clearContext();
```
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/dataelement/DataElementDefaultDimensionPopulator.java:135]: This feels like code that does exist elsewhere. I get we need this as `static` so if we have not yet created such helpers as static methods these should be put somewhere so we can reuse this code

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/startup/ConfigurationPopulator.java [hunk 8]
```diff
@@ -74,4 +90,20 @@ private void checkSecurityConfiguration() {
...
+    context.setAuthentication(authentication);
+    SecurityContextHolder.setContext(context);
+  }
+
+  private static void clearContext() {
+    SecurityContext context = SecurityContextHolder.getContext();
+    if (context != null) {
+      SecurityContextHolder.getContext().setAuthentication(null);
+    }
+    SecurityContextHolder.clearContext();
```
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/startup/ConfigurationPopulator.java:107]: Indeed, here is at least one more usage of the same code. I think this is also similar or the same as I added in the `DefaultAuthenticationService` so even that might want to call the same static helper methods.

File: dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/scheduling/RecordingJobProgressTest.java [hunk 9]
```diff
@@ -53,13 +56,19 @@
  *
  * @author Jan Bernitt
  */
-class RecordingJobProgressTest {
+class RecordingJobProgressTest extends TestBase {
```
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/scheduling/RecordingJobProgressTest.java:59]: This is a pure JUnit test, for these I would also prefer we don't mix in a base class that suggests spring when it is not involved. We rather should call a static method that is directly imported.

File: dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/deduplication/DeduplicationServiceTest.java [hunk 10]
```diff
@@ -163,6 +168,13 @@ void shouldBeAutoMergeable()
           PotentialDuplicateForbiddenException,
           ForbiddenException,
           NotFoundException {
+    SystemUser actingUser = new SystemUser();
+    Authentication authentication =
+        new UsernamePasswordAuthenticationToken(actingUser, "", actingUser.getAuthorities());
+    SecurityContext context = SecurityContextHolder.createEmptyContext();
+    context.setAuthentication(authentication);
+    SecurityContextHolder.setContext(context);
```
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/deduplication/DeduplicationServiceTest.java:176]: These lines look familiar :D I think this is the static helper again I pointed out or one that also should exist.

File: dhis-2/dhis-services/dhis-service-validation/src/main/java/org/hisp/dhis/validation/hibernate/HibernateValidationResultStore.java [hunk 11]
```diff
@@ -290,11 +290,11 @@ private String getQueryRestrictions(ValidationResultQuery query) {
    * @return String to add restrictions to the HQL query.
    */
   private String getUserRestrictions(SqlHelper sqlHelper) {
-    UserDetails currentUserDetails = CurrentUserUtil.getCurrentUserDetails();
-    if (currentUserDetails == null || currentUserDetails.isSuper()) {
+    //    !CurrentUserUtil.hasCurrentUser() ||
```
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-validation/src/main/java/org/hisp/dhis/validation/hibernate/HibernateValidationResultStore.java:293]: remove?

File: dhis-2/dhis-services/dhis-service-validation/src/test/java/org/hisp/dhis/validation/ValidationResultStoreHqlTest.java [hunk 12]
```diff
@@ -79,7 +81,7 @@
  *
  * @author Jan Bernitt
  */
-class ValidationResultStoreHqlTest {
+class ValidationResultStoreHqlTest extends TestBase {
```
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-validation/src/test/java/org/hisp/dhis/validation/ValidationResultStoreHqlTest.java:84]: Again, plain JUnit tests should not extend a base but import static methods

File: dhis-2/dhis-web-api/src/test/java/org/hisp/dhis/webapi/controller/tracker/export/event/EventRequestParamsMapperTest.java [hunk 13]
```diff
@@ -134,7 +133,7 @@ class EventRequestParamsMapperTest {
   public void setUp() throws ForbiddenException, NotFoundException, BadRequestException {
     User user = new User();
 
-    when(userService.getUserByUsername(CurrentUserUtil.getCurrentUsername())).thenReturn(user);
+    when(userService.getUserByUsername(null)).thenReturn(user);
```
Comment (jbee) [inline dhis-2/dhis-web-api/src/test/java/org/hisp/dhis/webapi/controller/tracker/export/event/EventRequestParamsMapperTest.java:136]: This is surprising - it clearly is not what it was - is this intentional?

**Review by jbee — CHANGES_REQUESTED** [review]
This looks really good. 
Very nicely removes a lot of annoying null checks.
Thanks for adding all the `@Nonnull` annotations.

It does need some cleanup with extracting some static helpers for reuse and importing static methods in plain JUnit and mockito tests instead of inheriting a base class.

Our `TestBase` has just too much spring-iness in it that is very confusing in the context of mockito or plain JUnit tests. I would like to avoid dragging in those details as they would need to be considered and deemed irrelevant. Also some stuff in our base class might not even work when not being in a spring based test.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/BaseIdentifiableObject.java (hunk 1 above)
Comment (jbee) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/BaseIdentifiableObject.java:427]: I mean make a `if (favorites == null) return false`

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/dataapproval/hibernate/HibernateDataApprovalStore.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (netroms) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/dataapproval/hibernate/HibernateDataApprovalStore.java:287]: This whole part should be refactored to use UserDetails later

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/user/CurrentUserUtil.java [hunk 14]
```diff
@@ -74,12 +77,13 @@ public static String getCurrentUsername() {
...
    *     unauthenticated
    */
+  @Nonnull
   public static UserDetails getCurrentUserDetails() {
     Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
     if (authentication == null
         || !authentication.isAuthenticated()
         || authentication.getPrincipal() == null) {
-      return null;
+      throw new RuntimeException("No authenticated user found");
```
Comment (jbee) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/user/CurrentUserUtil.java:86]: you missed this one to replace with `IllegalStateException`

File: dhis-2/dhis-test-web-api/src/test/java/org/hisp/dhis/webapi/controller/FileResourceControllerMockTest.java [hunk 15]
```diff
@@ -37,18 +37,21 @@
...
+import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
 import org.mockito.Mock;
 import org.mockito.junit.jupiter.MockitoExtension;
 import org.springframework.mock.web.MockHttpServletResponse;
 
 @ExtendWith(MockitoExtension.class)
-class FileResourceControllerMockTest {
+class FileResourceControllerMockTest extends TestBase {
```
Comment (jbee) [inline dhis-2/dhis-test-web-api/src/test/java/org/hisp/dhis/webapi/controller/FileResourceControllerMockTest.java:54]: Missed this one to not extend

File: dhis-2/dhis-web-api/src/test/java/org/hisp/dhis/webapi/controller/SharingControllerTest.java [hunk 16]
```diff
@@ -53,7 +56,7 @@
  * @author Volker Schmidt
  */
 @ExtendWith(MockitoExtension.class)
-class SharingControllerTest {
+class SharingControllerTest extends TestBase {
```
Comment (jbee) [inline dhis-2/dhis-web-api/src/test/java/org/hisp/dhis/webapi/controller/SharingControllerTest.java:59]: and this

File: dhis-2/dhis-web-api/src/test/java/org/hisp/dhis/webapi/controller/dataelement/DataElementOperandControllerTest.java [hunk 17]
```diff
@@ -97,7 +99,7 @@
  */
 @MockitoSettings(strictness = Strictness.LENIENT)
 @ExtendWith(MockitoExtension.class)
-class DataElementOperandControllerTest {
+class DataElementOperandControllerTest extends TestBase {
```
Comment (jbee) [inline dhis-2/dhis-web-api/src/test/java/org/hisp/dhis/webapi/controller/dataelement/DataElementOperandControllerTest.java:102]: and some more, maybe do a search ;)

File: dhis-2/dhis-web-api/src/test/java/org/hisp/dhis/webapi/service/GeoFeatureServiceMockTest.java [hunk 18]
```diff
@@ -156,4 +169,12 @@ private OrganisationUnit createOrgUnitWithCoordinates() throws IOException {
     ou.setGeometry(new GeometryJSON().read(POINT));
     return ou;
   }
+
+  private static void injectUserInSecurityContext(UserDetails user) {
```
Comment (jbee) [inline dhis-2/dhis-web-api/src/test/java/org/hisp/dhis/webapi/service/GeoFeatureServiceMockTest.java:173]: I think you now have a static helper for this or is it different?

**Review by jbee — APPROVED** [review]
You missed a couple but I approve assuming you change those as well.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/user/CurrentUserUtil.java [hunk 19]
```diff
@@ -177,15 +170,11 @@ public static void setUserSetting(UserSettingKey key, Serializable value) {
...
-          userSettings.put(key, value);
-        } else {
-          userSettings.remove(key);
-        }
-      }
+    Map<String, Serializable> userSettings = currentUser.getUserSettings();
+    if (value != null) {
+      userSettings.put(key, value);
+    } else {
+      userSettings.remove(key);
```
Comment (david-mackessy) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/user/CurrentUserUtil.java:177]: I know this is not part of this PR but seeing this seems a little odd.
The method's intention is to set a user setting, but if the value passed is null we remove the key. This sounds like a hidden side effect.

**Review by david-mackessy — APPROVED** [review]
great to see this

**Pull request merged**